### ISSUE TITLE
[Contribution] Layout configuration for a more usable font

### ISSUE THREAD

**Issue by zhguyu** [OpenNefia/OpenNefia#174]
Hi, I've been following this project for a while, and recently out of personal interest I'm trying to port the Chinese translation of Elona (by a group of volunteers from the Chinese community whose translation of ElonaPlus has been the de facto choice for the entire community and who has become the official partner of Elin by Noa) to this promising platform. Whether I can eventually make the port or not depends on the willingness of those people, but before that I'm currently working on an overhaul of font that would become the prerequisite for the translation port.
The reason is very simple: the font currently used (kochi gothic) doesn't support the entire CJK character set. For that reason, I'm working on migrating the font system of the project to Adobe Source Han Sans, which is a free font for commercial use, and very likely the most popular one in CJK regions.
Link: https://github.com/adobe-fonts/source-han-sans
This work involves tweaking the size and layout definitions of each UI element in the game, and my ultimate goal is to achieve pixel-wise parity between the new and old look.
Demonstration of some work-in-progress interfaces:
![Visual Studio-2024-01-09-22-51-22](https://github.com/OpenNefia/OpenNefia/assets/23335105/4e3212cc-d943-4b9e-abd5-2798f7118053)
![Visual Studio-2024-01-09-22-52-14](https://github.com/OpenNefia/OpenNefia/assets/23335105/7d9eda76-0988-4ddd-b410-0531caea3a62)

My question is: will you be accepting such pull request when the above work is finished? If not I will just keep them in my fork. Also, please do tell me if you have other requirements for font, such as whether it should be strictly pixelated or monospaced, etc. Thank you.

**Comment by Ruin0x11** [comment]
Hello and thank you for reaching out. I do want to support the font you selected since ideally *any* CJK font should be usable in the UI, and I think this is a good opportunity for fixing font spacing issues.

Originally I was aiming to aim for pixel parity with MS Gothic since that was 1.22's default CJK font. I would like to take a look at what needs to be changed in your PR; ideally the user should be able to use either MS Gothic or Source Han Sans without the UI looking strange. Potentially I could adjust some of the font-related bits with a padding/spacing config option specific to each font. I think even 1.22 had a few options for adjusting font spacing for a few UI locations. Hopefully those options can be applied relatively uniformly throughout the UI so far.

**Comment by zhguyu** [comment]
Thank you for the reply. I'll send the pull request shortly.

**Comment by Ruin0x11** [comment]
Some more information about the status of the port in case you're curious: currently I'm in the middle of porting all of the spell effects, specifically the buff system. It's basically going at the pace of whenever I have some free time and motivation to work on it a bit more, so slow and steady. However, at one point I did manage to port almost all of the game in the past, so I have a bit of knowledge as to how the internals should generally look by now. There is another TODO file in the repo root with a rough estimate of the work still needing to be done.

Feel free to reach out if you have any questions about the codebase or design and I'd be happy to answer.

**Comment by zhguyu** [comment]
I see. Currently I'm working on tweaking the overall layout of the game to make it look better with Source Sans. This will break the pixel parity with 1.22 but I think it's not such a big deal after all since vanilla's look have it's own problems. For example, I can see that you make patches here and there to replicate the texts not aligned to each other horizontally in vanilla, but on the other hand, OpenNefia is a new platform where UI changes are commonly expectable. Thus I don't think it would be a problem if the changes are towards the better.
This work includes relocating not only texts but also other assets, such as the keyhint text and icon in the second image are adjusted to be in the middle between the line above the the window border below.
![Visual Studio-2024-01-12-16-44-43](https://github.com/OpenNefia/OpenNefia/assets/23335105/7f20e5a5-5f94-4be2-8042-9918d1fe364f)
![Visual Studio-2024-01-12-16-45-32](https://github.com/OpenNefia/OpenNefia/assets/23335105/acb522ed-03a6-4995-a6e0-4dd2e32d3075)

**Comment by Ruin0x11** [comment]
I think so long as the changes are negligible it should be okay; I'd want to see where new offsets are necessary. The main issue I can imagine would be text slightly overshooting the boundaries of the windows or crashing into other UI elements.

**Comment by zhguyu** [comment]
Absolutely. This will be a larger piece of work and will be in its own PR.

**Comment by Ruin0x11** [comment]
Also if you or the others would like to provide the Chinese translation, I would be happy to have it merged into the main codebase. It would make the port even more accessible to others. I can provide support for the localization system in that case (it supports custom functions for things like pronouns).

**Comment by zhguyu** [comment]
The translation is definitely on my list of things to do, but probably won't happen in near future as the team is focused on Elin at the moment and also E+ 2.21fix2. Most likely I can communicate with them after either of them is finished.

I think it's a good idea to distinguish between QoL features, balance changes and actual contents. I agree that recursive forking plus a decade of development would eventually make the codebase hard to maintain or work with, as has been the case for Plus or oomSEST or practically any famous Elona variant. My concern is will it be possible that different mods be loaded simultaneously and any conflict be resolved by the engine?

**Comment by Ruin0x11** [comment]
By far the modding will be the trickiest part to get right. It's kind of the wild west since nothing like it has been attempted for Elona, and it's somewhat difficult to find open-source examples of engines that accomplish this.

I never got as far as porting a lot of content for Plus or omake, but my impression is people will want to tweak parts of the game balance. The best idea I have in mind so far is to modularize the code such that the relevant formulas are exposed as methods. Then an assembly patching library like [Harmony](https://github.com/pardeike/Harmony) can target those methods and change their behavior. This is how a lot of Unity games support modding and it seems pretty flexible.

This alone is fine, but for Elona's case there are conflicts between how each variant handles game balance. Each variant provides their own interpretation of each method implementation, so to speak, and they can potentially conflict with each other. At the same time, it would be nice to say "install the Plus and omake content, take these bits of the game I like, and leave out the rest" without having to mess around at the mod dependency layer too much. So my idea for that would be allowing each mod to provide their game balance tweaks as a "[pointcut](https://en.wikipedia.org/wiki/Pointcut)" that can be switched between with a config option. Again, I have no idea how this will be implemented in practice since it's just an idea in my head right now. The best solution will probably be reached at the end of a lot of prototyping and failed experiments.

As for other potential areas of conflicts... there will probably be a ton. Including a bunch of unknown ones that will only become apparent later. Not to mention, how will the save data added by mods be managed and migrated when mod versions change? It's a complex problem overall.
